### PR TITLE
[transformer] add multi warmup and learning rate for different modules

### PR DESCRIPTION
**Pull request by Mddct**
当微调的时候，模型的不同部分需要的学习率和warmup不一样,比如以下 微调whisper的手 因为没有ctc的weight

我门可以设置ctc的warmu up为12000， lr 0.01， 其他部分为1200,0.00001
```yaml
optim: adam
optim_conf:
  lr: [0.01, 0.00001]
  modules: ['ctc']
scheduler: warmuplr
scheduler_conf:
  warmup_steps: [1500,5000]
```
### TODO
- [x] whisper aishell result
- [x] tensorboard multiple lr
- [x] refactor code and rebase main

### CODE REVIEW

**Comment by Mddct** [comment]
截屏2024-03-28 20 09 28

**Comment by Mddct** [comment]
tensorboard works！
截屏2024-03-28 22 56 02

**Comment by Mddct** [comment]
单个lrr+warmup ， cv和train work
截屏2024-03-29 11 25 15

**Comment by Mddct** [comment]
https://github.com/wenet-e2e/wenet/pull/2449/commits/10b704570bf2e0b762ab1db715adc895c0464a2d 这个commit 重构了cv 和train 以及 lr 的打印逻辑

单个lr+warmup ， cv和train work

image

**Comment by Mddct** [comment]
(TBD)
===========================================================================
ctc_greedy_search
Overall -> 6.54 % N=104765 C=98887 S=5695 D=183 I=978
Mandarin -> 5.94 % N=104762 C=98887 S=5692 D=183 I=344
Other -> 21233.33 % N=3 C=0 S=3 D=0 I=634

**Comment by fclearner** [comment]
> ## fsdp (TBD)
> ### 8卡A100， step 模式 ,3000 save interval, **fp32**, **zero2**
> * Feature info: using log_mel_spectrogram feature, no cmvn, no speed perturb
> * Training info: conf/finetune_whisper_largev3_conv2d4.yaml
> * Decoding info: ctc_weight 0.3, average_num 3
> * Git hash: TBD
> 
> decoding mode	CER
> attention decoder	3.09 % N=104765 C=101638 S=3024 D=103 I=108
> ctc greedy search	6.24 % N=104765 C=99161 S=5422 D=182 I=936
> ctc prefix beam search	6.18 % N=104765 C=99158 S=5429 D=178 I=863
> attention rescoring	4.64 % N=104765 C=100207 S=4452 D=106 I=301

强，ctc的收益是因为使用了不同的学习率吗

**Comment by Mddct** [comment]
> > ## fsdp (TBD)
> > ### 8卡A100， step 模式 ,3000 save interval, **fp32**, **zero2**
> > 
> > * Feature info: using log_mel_spectrogram feature, no cmvn, no speed perturb
> > * Training info: conf/finetune_whisper_largev3_conv2d4.yaml
> > * Decoding info: ctc_weight 0.3, average_num 3
> > * Git hash: TBD
> > 
> > decoding mode	CER
> > attention decoder	3.09 % N=104765 C=101638 S=3024 D=103 I=108
> > ctc greedy search	6.24 % N=104765 C=99161 S=5422 D=182 I=936
> > ctc prefix beam search	6.18 % N=104765 C=99158 S=5429 D=178 I=863
> > attention rescoring	4.64 % N=104765 C=100207 S=4452 D=106 I=301
> 
> 强，ctc的收益是因为使用了不同的学习率吗

直觉上，整个模型是pretrain的 除了ctc和conv2d， 所以只让这两个较大的学习率 ‘快速的学习’，其他的stable training
可以参考 bestrq paper： https://arxiv.org/pdf/2202.01855.pdf
101551711637944_ pic

未来也会支持w2vbert的fintune 所以这里就引入了

**Comment by Mddct** [comment]
https://github.com/wenet-e2e/wenet/pull/2412 修复了A100 和v100精度问题， 这里实验需要重新跑下 （预期还会更好些）

**Comment by Mddct** [comment]
https://github.com/wenet-e2e/wenet/pull/2412 使用修复后的的结果：
##  8卡A100， step 模式 ,3000 save interval, **bf16**, **zero2** 
| decoding mode             | CER   |
|---------------------------|-------|
| attention decoder         |  3.13 % N=104765 C=101598 S=3068 D=99 I=111 ｜
| ctc greedy search         | 6.10 % N=104765 C=99224 S=5366 D=175 I=845 |
| ctc prefix beam search    | 6.03 % N=104765 C=99223 S=5367 D=175 I=776 |
| attention rescoring       | 4.52 % N=104765 C=100241 S=4420 D=104 I=215 |

**Comment by Mddct** [comment]
```yaml
optim: adam
optim_conf:
  lr: [0.001, 0.00005, 0.00001]
  modules: ['ctc', 'encoder.embed']
scheduler: warmuplr
scheduler_conf:
  warmup_steps: [1500, 10000, 5000]
```
##  8卡A100， step 模式 ,3000 save interval, **bf16**, **zero2** 
| decoding mode             | CER   |
|---------------------------|-------|
| attention decoder         |  3.09 % N=104765 C=101626 S=3039 D=100 I=103｜
| ctc greedy search         | 6.04 % N=104765 C=99335 S=5266 D=164 I=899 |
| ctc prefix beam search    | 5.96 % N=104765 C=99345 S=5258 D=162 I=821 |
| attention rescoring       |  4.47 % N=104765 C=100289 S=4370 D=106 I=212|

**Comment by Mddct** [comment]
image
代表我们已经跑了0个epoch300个step，保存模型299.pt  （从0.pt开始）

File: wenet/utils/train_utils.py [hunk 1]
```diff
@@ -439,10 +439,38 @@ def wrap_cuda_model(args, model, configs=None):
...
+            if id(param) not in special_param_ids:
+                rest_params.append(param)
+        groups.append({'params': rest_params, 'lr': lr[-1]})
+
+    params = groups if len(groups) > 0 else model.parameters()
+    optim_conf = copy.deepcopy(configs['optim_conf'])
+    if 'modules' in optim_conf:
+        del optim_conf['modules']
+    if isinstance(lr, List):
+        optim_conf['lr'] = lr[-1]
```
Comment (xingchensong) [inline wenet/utils/train_utils.py:469]: 这一步是为啥，optimizer必须传一个lr参数且不能是list？

**Review by xingchensong — APPROVED** [review]
